File: avl.py
from graphviz import Graph

class Avl:
    def __init__(avl, nom):
        avl.nom = nom
        avl.fils_gauche = None
        avl.fils_droit = None
        avl.hauteur = 1

    def feuille(self):
        return self.fils_gauche is None and self.fils_droit is None

def liste_aretes(avl):
    liste = []
    if avl.fils_gauche is not None:
        liste.append((str(avl.nom), str(avl.fils_gauche.nom)))
        liste += liste_aretes(avl.fils_gauche)
    if avl.fils_droit is not None:
        liste.append((str(avl.nom), str(avl.fils_droit.nom) ))
        liste += liste_aretes(avl.fils_droit)
    return liste

def affiche(avl):
    graph = Graph()
    graph.edges(liste_aretes(avl))
    graph.render(view=True)
        
def rotation_gauche(avl):
    b = avl.fils_gauche
    avl.fils_gauche = b.fils_droit
    maj(avl)
    b.fils_droit = avl
    maj(b)
    return b

def rotation_droite(avl):
    c = avl.fils_droit
    avl.fils_droit = c.fils_gauche
    maj(avl)
    c.fils_gauche = avl
    maj(c)
    return c

def hauteur(avl):
    return 0 if avl is None else avl.hauteur

def dico_hauteur(avl):
    if avl is None:
        return {}
    dico = {avl.nom: avl.hauteur}
    dico.update(dico_hauteur(avl.fils_gauche))
    dico.update(dico_hauteur(avl.fils_droit))
    return dico

def balance(avl):
    return hauteur(avl.fils_droit) - hauteur(avl.fils_gauche)

def maj(avl):
    avl.hauteur = 1 + max(hauteur(avl.fils_droit), hauteur(avl.fils_gauche))

def equilibre(avl):
    if balance(avl) == -2:
        if balance(avl.fils_gauche) > 0:
            avl.fils_gauche = rotation_droite(avl.fils_gauche)
        avl = rotation_gauche(avl)
    if balance(avl) == 2:
        if balance(avl.fils_droit) <0:
            avl.fils_droit = rotation_gauche(avl.fils_droit)
        avl = rotation_droite(avl)
    maj(avl)
    return avl

def insert(avl, valeur):
    if avl is None:
        return Avl(valeur)
    if valeur == avl.nom:
        return avl
    if valeur < avl.nom:
        avl.fils_gauche = insert(avl.fils_gauche, valeur)
    else:
        avl.fils_droit = insert(avl.fils_droit, valeur)
    maj(avl)
    return equilibre(avl)

def recherche(avl, valeur):
    "renvoie le sous arbre ayant valeur pour racine. Si il n'existe pas renvoie None"
    if avl is None:
        return avl
    if valeur == avl.nom:
        return avl
    if valeur < avl.nom:
        return recherche(avl.fils_gauche, valeur)
    else:
        return recherche(avl.fils_droit, valeur)

def successeur(avl):
    "renvoie le successeur inorder de la racine, None si il n'y a pas de fils droit"
    if avl.fils_droit is None:
        return None
    succ = avl.fils_droit
    while succ.fils_gauche is not None:
        succ = succ.fils_gauche
    return succ


def supprime(avl, valeur):
    if avl is None:
        return None
    if valeur < avl.nom:
        avl.fils_gauche = supprime(avl.fils_gauche, valeur)
    elif valeur > avl.nom:
        avl.fils_droit = supprime(avl.fils_droit, valeur)
    else:
        if avl.feuille():
            return None
        elif avl.fils_droit is None:
            return avl.fils_gauche
        elif avl.fils_gauche is None:
            return avl.fils_droit
        else:
            succ = successeur(avl)
            avl.nom = succ.nom
            avl.fils_droit = supprime(avl.fils_droit, avl.nom)
    maj(avl)
 
    return equilibre(avl)


avl = Avl(0)
for i in range(1, 50):
    avl = insert(avl, i)
affiche(avl)
input()
for i in range(1, 50,):
    print(i)
    avl = supprime(avl, i)
    affiche(avl)
    input()

File: avl.c
#include <stdio.h>
#include <stdlib.h>

struct Noeud {
    int val ;
    struct Noeud * gauche ;
    struct Noeud * droit ;
    int hauteur;
};

void infixe(struct Noeud * root) {
    int * parcours;
    if (root->gauche != NULL) {
        infixe(root->gauche);
    }
    printf("%i ( %i ) ", root->val, root->hauteur);
    if (root->droit != NULL) {
        infixe(root->droit);
    }
}

struct Noeud * clean(struct Noeud * root) {
    if (root->gauche != NULL) {
        clean(root->gauche);
    }
    if (root->droit != NULL) {
        clean(root->droit);
    }
    free(root);
    return NULL;
}


int hauteur(struct Noeud * noeud) {
    return (noeud == NULL ? 0: noeud->hauteur);
}

void maj(struct Noeud * noeud) {
    if (hauteur(noeud->gauche) > hauteur(noeud->droit)) {
        noeud->hauteur = 1 + hauteur(noeud->gauche);
    }
    else {
        noeud->hauteur = 1 + hauteur(noeud->droit);
    }
}


int balance(struct Noeud * noeud) {
    return hauteur(noeud->droit) - hauteur(noeud->gauche);
}


struct Noeud * rotation_droite(struct Noeud * noeud) {
    if (noeud->droit == NULL) return noeud;
    struct Noeud * tmp = noeud->droit;
    noeud->droit = tmp->gauche;
    maj(noeud);
    tmp->gauche = noeud;
    maj(tmp);
    return tmp;
}

struct Noeud * rotation_gauche(struct Noeud * noeud) {
    if (noeud->gauche == NULL) return noeud;
    struct Noeud * tmp = noeud->gauche;
    noeud->gauche = tmp->droit;
    maj(noeud);
    tmp->droit = noeud;
    maj(tmp);
    return tmp;
}


struct Noeud * equilibre(struct Noeud * noeud) {
    int b = balance(noeud);
    //printf("noeud %i hauteur %i balance %i \n", noeud->val, hauteur(noeud), b);
    if (b == -2) {
        if (balance(noeud->gauche) > 0) {
            noeud->gauche = rotation_droite(noeud->gauche);
            maj(noeud->gauche);
        }
        noeud = rotation_gauche(noeud);
        maj(noeud);
    }
    if (b == 2) {
        if (balance(noeud->droit) < 0) {
            noeud->droit = rotation_gauche(noeud->droit);
            maj(noeud->droit);
        }
        noeud = rotation_droite(noeud);
        maj(noeud);
    }
    return noeud;
}

struct Noeud * insere(struct Noeud * root, int val) {
    if (root == NULL) {
        struct Noeud * feuille = (struct Noeud *) malloc(sizeof(struct Noeud));
        feuille->val = val;
        feuille->gauche = NULL;
        feuille->droit = NULL;
        feuille->hauteur = 1;
        return feuille;
    }
    if (val < root->val) {
        root->gauche = insere(root->gauche, val);
    }
    else if (val > root->val) {
        root->droit = insere(root->droit, val);
    }
    maj(root);
    root = equilibre(root);
    return root;
}

int successeur(struct Noeud * root) {
    struct Noeud * noeud = root;
    while (noeud->gauche != NULL) noeud = noeud->gauche;
    return noeud->val;
}

struct Noeud * supprime(struct Noeud * root, int val) {
    if (val < root->val) {
        root->gauche = supprime(root->gauche, val);
        printf("supprime à gauche root = %d\n", root->val);
        infixe(root);
        printf("\n");
        maj(root);
        return equilibre(root);
    }
    else {
        if (val > root->val) {
        root->droit = supprime(root->droit, val);
        printf("supprime à droite root = %d\n", root->val);

        infixe(root);
            printf("\n");
        maj(root);
        return equilibre(root);
        }   
        else {
            //root->val = val : Le noeud à supprimer
            printf("supprime  %d\n", root->val);

            if ((root->gauche == NULL) & (root->droit == NULL)) {
                printf("feuille\n");
                free(root);
                return NULL;      //feuille
            }
            if (root->gauche == NULL) {
                printf("pas de gauche\n");
                struct Noeud * tmp = root->droit;
                free(root);
                return tmp;
            }   //un seul fils droit
            if (root->droit == NULL) {
                printf("pas de droit\n");
                struct Noeud * tmp = root->gauche;
                free(root);
                return tmp;
            }    //un seul fils gauche
            //cas général
            printf("cas general\n");
            int succ = successeur(root);
            root->val = succ;
            root->droit = supprime(root->droit, succ);
            printf("succ = %d\n", succ);
            infixe(root);
            printf("\n");
            maj(root);
            return equilibre(root);
            }
        }
    }


int main() {
    struct Noeud * root = (struct Noeud *) malloc(sizeof(struct Noeud));
    root->val = 0;
    root->gauche = NULL;
    root->droit = NULL;
    for (int i = 1; i < 8; i++) {
        root = insere(root, i);
    }
    infixe(root);
    printf("balance %i\n", balance(root));
    for (int i = 1; i < 2; i++) {
        printf("i = %i\n", i);
        root = supprime(root, i);
        infixe(root);
        printf("\n");
        printf("balance %i\n", balance(root));
    }
    
    clean(root);
    //free(root);
    return 0;
}
